src/main/scala/com/microsoft/ml/spark/lightgbm/LightGBMBase.scala

```diff
@@ -39,10 +41,38 @@ trait LightGBMBase[TrainedModel <: Model[TrainedModel]] extends Estimator[Traine
     }
   }
 
-  protected def prepareDataframe(dataset: Dataset[_], trainingCols: ListBuffer[String],
+  protected def castColumns(dataset: Dataset[_], trainingCols: ListBuffer[(String, DataType)],
+                            selectedCols: ListBuffer[Column]): DataFrame = {
+    val schema = dataset.schema
+    // Figure out which columns need to be cast
+    val castCols = trainingCols.flatMap {
+      case (name, datatype) => {
+        val index = schema.fieldIndex(name)
+        // Note: We only want to cast if original column was of numeric type
+        schema.fields(index).dataType match {
+          case numericDataType: NumericType =>
+            if (numericDataType != datatype) Some(name, datatype)
+            else None
+          case _ => None
+        }
+      }
+    }.toMap
+    // Cast columns to correct types
+    dataset.select(
+      selectedCols.map { selectedCol =>
+        val selectedColName = selectedCol.toString
+        if (castCols.contains(selectedColName)) dataset(selectedColName).cast(castCols(selectedColName))
+        else dataset(selectedColName)
+      }: _*
+    ).toDF()
+  }
+
+  protected def prepareDataframe(dataset: Dataset[_], trainingCols: ListBuffer[(String, DataType)],
                                  numWorkers: Int): DataFrame = {
+    // Select only relevant columns
+    val selectedCols = trainingCols.map { case (name, _) => col(name) }
+    val df = castColumns(dataset, trainingCols, selectedCols)
```

Comment:
if theres no diff between training and selected cols, move the selected cols computation inside the function and simplify please

Reply:
ah good catch, done!